NFC shell (#11790)
nrfconnect/sdk-nrf: doc/nrf/libraries/shell/shell_nfc.rst

.. _shell_nfc_readme:

NFC shell transport
###################

.. contents::
   :local:
   :depth: 2

The NFC shell transport allows you to receive shell commands remotely over the ISO-DEP (ISO14443-4A) protocol with any polling NFC device.
It uses the :ref:`type_4_tag` library.

The NFC shell transport is used by default in the :ref:`nfc_shell` sample.

Overview
********

The NFC shell transport uses the NFC T4T library to emulate the NFC Tag device.

> The NFC Tag is a passive device that sends response to the Reader/Writer device payload.

Author: I would keep here Reader/Writer or polling device

> The NFC Tag is a passive device that sends response to the Reader/Writer device payload.

Reviewer: polling device will be ok :)

It may happen that one or both devices do not have any data to exchange.
The NFC tag always responds to the polling device's frame.
A record payload always starts with a dummy one-byte header.
Its value is not relevant and it is not checked by this library.
If one of the devices does not have any data, then only the header byte is sent.

The following diagram illustrates various data exchange cases:

.. msc::
    hscale = "1.3";
    a [label="Reader/Writer"], b [label="Tag"];
    |||;
    a rbox b [label = "Empty packet exchange(poll for data)"];
    a=>b [label = "data: ['S'] - one dummy byte, 'S' character here"];
    a<=b [label = "data: ['S'] - one dummy byte"];
    |||;
    a rbox b  [label = "Only Reader\Writer has data"];
    a=>b [label = "data: ['S', 1..N bytes]"];
    a<=b [label = "data: ['S'] - one dummy byte"];
    |||;
    a rbox b [label = "Only Reader\Writer has data"];
    a=>b     [label = "data: ['S'] - one dummy byte"];
    a<=b     [label = "data: ['S', 1..N bytes]"];
    |||;
    a rbox b [label = "Both devices have data"];
    a=>b [label = "data: ['S', 1..N bytes]"];
    a<=b [label = "data: ['S', 1..N bytes]"];

Requirements
************

> This library has the following requirements:

Author: Discussed offline

  * The :kconfig:option:`CONFIG_NFC_THREAD_CALLBACK` must be enabled.
    The NFC callback function must be called from a thread context.
    If default configuration it is called from the system workqueue context.
    This library uses the operating system primitives in the callback function which cannot be
    used in an interrupt context.
  * If library logs and the logger are enabled, then the logger should not be used in the synchronous mode.
    Synchronous logs processing can introduce to big latency in NFC events processing which causing the NFC shell backend being unstable.
    If you need logs from this library, configure logger to be in the deferred mode by setting the :kconfig:option:`CONFIG_LOG_MODE_DEFERRED` Kconfig option.

Configuration
*************

Set the :kconfig:option:`CONFIG_SHELL_NFC` Kconfig option to enable the shell backend.

The following configuration options are available for this module:

   * :kconfig:option:`CONFIG_SHELL_NFC_INIT_PRIORITY` sets initialization priority for the NFC shell backend.
   * :kconfig:option:`CONFIG_SHELL_NFC_BACKEND_TX_BUFFER` sets the NFC T4T APDU response buffer size.

>      It size might depends on your Reader/Writer maxiumum supported APDU size.

Author: Ok I renamed it in documentation and in the code

   * :kconfig:option:`CONFIG_SHELL_NFC_BACKEND_TX_RING_BUFFER_SIZE` sets the shell Tx ring buffer size.

>      It is used to buffer data coming from shell interface for the NFC T4T transport, that can send data only on the data reception from the Reader/Writer.

Reviewer: ```suggestion
     It is used to buffer data from the shell interface through the NFC T4T transport, that can send data only on the data reception from the polling device.
```
Could you please clarify the second part of the sentence? 
1. **What** can send data?
2. What is **sending data on the data reception**?

Author: !. Transport, data is buffered because it cannot be send right after shell request it. Library need to wait for data reception from polling device and sends s

     It handles shell incoming data from the NFC transport.

See the Kconfig help for details.

Usage
*****

You can set the :kconfig:option:`CONFIG_SHELL_NFC` Kconfig option to enable the NFC shell.
The shell is initialized automatically during the system initialization.
You can approach your polling device that uses the raw ISO-DEP (ISO14443-4A) protocol to start the shell interface.
Next, forward the received data to the terminal and send back data received from terminal through your polling device.

> For example, it can be the keyboard data or the bulk data.

Author: Discussed offline

For example, it can be the keyboard data or the bulk data.

.. note::
   The NFC transport can be physically disconnected and any time. After disconnection all data sent by the shell will be discarded.
   The incomplete commands sent by the Reader/Writer before transport lost are discarded also.
   Every transport connection or re-connection starts a new shell session.

Samples using the library
*************************

The following |NCS| samples use this library:

* :ref:`nfc_shell`

Limitations
***********

The NFC T4T ISO-DEP transport is relatively slow.
The maximum bit rates is 106 kbits/s for NFC-A technology.

> Data is sent in blocks and Reader/Writer device needs to send data first or send dummy data to poll data from the Tag.

Reviewer: ```suggestion
Data is sent in blocks.
The polling device needs to send data first or to send dummy data to poll data from the tag.
```
What is **poll data**?
Why we need info that data is sent in blocks?
Are they **blocks** indeed, or do you mean, e.g., messages or records?

Author: I mean poll for data :) ISO-DEP sends data in blocks. This is described in its specification

> This cause that, human visible latency occurs during printing characters on the terminal.

Reviewer: ```suggestion
As a result, latency visible to the unaided eye occurs during printing characters in the terminal.
```

> It is not recommended also to set NFC shell as logger backend, due to transport speed.

Reviewer: ```suggestion
Moreover, it is not recommended to set the NFC shell as a logger backend due to the transport speed.
```

> You can set the :kconfig:option:`CONFIG_SHELL_NFC_INIT_LOG_LEVEL_NONE` Kconfig option that, set the initial log level to none for the NFC shell.

Reviewer: ```suggestion
In the :kconfig:option:`CONFIG_SHELL_NFC_INIT_LOG_LEVEL_NONE` Kconfig option you can set the initial log level to ``none`` for the NFC shell.
```

Author: I used your suggestion but I am not sure in 100 % here
